src/sv2_template_provider.h

```diff
@@ -0,0 +1,258 @@
+#ifndef BITCOIN_SV2_TEMPLATE_PROVIDER_H
+#define BITCOIN_SV2_TEMPLATE_PROVIDER_H
+
+#include <logging.h>
+#include <node/sv2_messages.h>
+#include <node/miner.h>
+#include <streams.h>
+#include <util/sock.h>
+
+class ChainstateManager;
+class CTxMemPool;
+
+struct Sv2Client
+{
+    /**
+     * Receiving and sending socket for the connected client
+     */
+    std::shared_ptr<Sock> m_sock;
+
+    /**
+     * Whether the client has confirmed the connection with a successful SetupConnection.
+     */
+    bool m_setup_connection_confirmed;
+
+    /**
+     * Whether the client is a candidate for disconnection.
+     */
+    bool m_disconnect_flag;
+
+    /**
+     * Whether the client has received CoinbaseOutputDataSize message.
+     */
+    bool m_coinbase_output_data_size_recv;
+
+    /**
+     * Specific additional coinbase tx output size required for the client.
+     */
+    unsigned int m_coinbase_tx_outputs_size;
+
+    explicit Sv2Client(std::shared_ptr<Sock> sock) : m_sock{sock} {};
+};
+
+struct Sv2TemplateProviderOptions
+{
+    /**
+     * The default listening port for the server.
+     */
+    uint16_t port = 8442;
+
+    /**
+     * The current protocol version of stratum v2 supported by the server. Not to be confused
+     * with byte value of identitying the stratum v2 subprotocol.
+     */
+    uint16_t protocol_version = 2;
+
+    /**
+     * Optional protocol features provided by the server.
+     */
+    uint16_t optional_features = 0;
+
+    /**
+     * The default option for the additional space required for coinbase output.
+     */
+    unsigned int default_coinbase_tx_additional_output_size = 0;
+
+    /**
+     * The default flag for all new work.
+     */
+    bool default_future_templates = true;
```

Comment:
This is true, I treated this PR as draft. Currently in development and testing against the SRI, all templates are assumed to be future so this would be updated if this PR gains interest for a complete implementation.